### PR TITLE
API: Support running gadgets from files and memory

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This PR changes some interfaces and functions to allow running a gadget from a file or from memory and implements two examples showing it.

### TODO
- [x] How to avoid the github.com/quay/claircore dependency being propagated to our go.mod

### CODE REVIEW

**Comment by flyth** [comment]
Should we also give the option to push the binary using gRPC?
I think that would be helpful especially for debugging purposes without having to go through external registries. It should be disabled by default server-side, though.

**Comment by alban** [comment]
> Should we also give the option to push the binary using gRPC? I think that would be helpful especially for debugging purposes without having to go through external registries.

Why not, but in a separate PR.

> It should be disabled by default server-side, though.

Why? I think options to enable/disable it is good. But I don't think it is necessarily less secure than downloading from a OCI registry. That depends on the context and if users verify the image signature.

**Comment by flyth** [comment]
> > Should we also give the option to push the binary using gRPC? I think that would be helpful especially for debugging purposes without having to go through external registries.
> 
> Why not, but in a separate PR.
> 
> > It should be disabled by default server-side, though.
> 
> Why? I think options to enable/disable it is good. But I don't think it is necessarily less secure than downloading from a OCI registry. That depends on the context and if users verify the image signature.

Agree with both, just wanted to start the discussion to keep it in mind when developing this further.

I also don't think it's less secure per se if it runs through all checks as well; still there are certain things we need to consider, which is why I'd suggest to keep it disabled by default:

If you'd for example bundle just a couple of core gadgets with your IG installation, you'd keep signature checking on (by default) and deploy it. You'd probably not want for IG to be able to run any other non-bundled (but signed) core gadget using the gRPC way, because it could leak private information or whatnot (you actively chose to not bundle it in the first place).

The more use cases we add, the more complex the security situation becomes (edge cases, unwanted behavior) - and thus IMHO we should always be more restrictive than permissive by default so that the admin actively has to enable features like this after careful consideration.

**Comment by mauriciovasquezbernal** [comment]
> Can you have another set of go.mod/go.sum in the examples/gadgets/simple/from_file/ and examples/gadgets/simple/from_memory/ sub-directories?


Added a single pair on examples/. I also needed to add a go.work file.

File: pkg/operators/oci-handler/oci.go [hunk 1]
```diff
@@ -188,21 +188,33 @@ func (o *OciHandlerInstance) init(gadgetCtx operators.GadgetContext) error {
...
+	// need to be sure the image is available.
+	if target == nil {
+		var err error
+		target, err = oci.GetLocalOciStore()
+		if err != nil {
+			return fmt.Errorf("getting local oci store: %w", err)
+		}
+
+		// Make sure the image is available, either through pulling or by just accessing a local copy
+		// TODO: add security constraints (e.g. don't allow pulling - add GlobalParams for that)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/operators/oci-handler/oci.go:202]: This comment was already there. Code was moved into a new conditional.

**Comment by mauriciovasquezbernal** [comment]
Marking as ready. I don't have anything against allowing something similar on the gRPC interface, but IMO it doesn't belong to this PR and the current implementation doesn't collides with that. Any thoughts?

File: pkg/oci/oci.go [hunk 2]
```diff
@@ -80,9 +80,9 @@ const (
 	localhost = "localhost"
 )
 
-// getLocalOciStore returns a single local oci store. oci.Store is concurrently safe only
+// GetLocalOciStore returns a single local oci store. oci.Store is concurrently safe only
 // against its own instance inside the same go program
-var getLocalOciStore = sync.OnceValues(func() (*oci.Store, error) {
+var GetLocalOciStore = sync.OnceValues(func() (*oci.Store, error) {
```
Comment (flyth) [inline pkg/oci/oci.go:85]: Maybe a function would be better than an exported var that could be changed on runtime.

File: pkg/operators/oci-handler/oci.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/operators/oci-handler/oci.go:202]: Added some prerequisites for that to #2901 - will follow up with a PR moving the settings to global once that is merged. Then we can tighten it even more.

**Review by flyth — APPROVED** [review]
Looks really good! Nice to have this possibility as well!

File: examples/gadgets/simple/from_file/main.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+func do() error {
+	ctx := context.Background()
+
+	// Create the OCI store from a tarball.
+	// To create this tarball run:
+	// $ sudo ig image pull trace_open:latest
+	// $ sudo ig image export trace_open:latest /tmp/trace_open.tar
+	ociStore, err := orasoci.NewFromTar(ctx, "trace_open.tar")
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply examples/gadgets/simple/from_file/main.go:38]: My idea is to create an example that can be compiled without taking any additional steps. I don't think there is another possibility than storing the gadget image in the repo (the same we're doing for built-in gadgets).

**Review by burak-ok — APPROVED** [review]
LGTM,

I had thought that this PR would additionally add a flag like `--from-file` to `ig run` to directly run from a tarball.
For `ig` there is no problem, but as written previously by others, for `kubectl-gadget` we would need to push the archive/eBPF program through gRPC

**Comment by mauriciovasquezbernal** [comment]
> I had thought that this PR would additionally add a flag like --from-file to ig run to directly run from a tarball.

I initially implemented it that way, but then I realized it doesn't make sense now that we have the image export and import commands. We can discuss later on how to handle this for the kubectl-gadget case. 

Thanks for the review.

**Pull request merged**